src/reportengine/configparser.py

```diff
@@ -132,6 +134,66 @@ def f_(self, val, *args, **kwargs):
 
     return f_
 
+def record_from_defaults(f):
+    """Decorator for recording default values. Given a key, for example
+    `filter_defaults` there might exist several specifications or specs of
+    defaults, each spec will have its own label e.g foo and bar. First a
+    parse function must be defined for the key and decorated with
+    `record_from_defaults`
+
+        @record_from_defaults
+        def parse_filter_defaults(self, spec):
+            return spec
+
+    Next a rule must exist for obtaining the defaults, the most simple rule
+    would check for a provided spec in a dictionary. These rules must be named
+    `load_default_<key>` e.g
+
+        def load_default_filter_defaults(self, spec):
```

Comment:
So this method should get called if the `filter_defaults` namespace _isn't_ in the runcard right @wilsonmr @Zaharid ?? For me it works the other way, this get's called if I _do_ put `filter_defaults` in the runcard. Perhaps im just confused.

Also, what's the difference between spec and key here

Reply:
so key is the think you want to record defaults for. Spec is the particular default.. Think that we will have a `"3.1"` default and a `"4.0"` default and in the future probably a `"4.1"` defaults. These are the specs. The load_default* function takes a spec, for example "3.1" and turns that into the actual defaults. So in your case the 3.1 default would be that expression without the `abs` and the "4.0" default would be with (if I got that the right way round), for example.

The load_default_filter_defaults will be called whenever parse_filter_defaults is called *unless* you also have a filter_defaults_recorded_spec_ in the namespace (because then it's assumed to be a lockfile). it causes the lockfile to be written with something like:

```yaml
filter_defaults_recorded_spec_:
 # spec: result
 "3.1": <whatever the 3.1 default actually is>
```

but obviously if that already exists as well then instead we do

`<whatever the default *was*> = filter_defaults_recorded_spec_[spec]`

which means if the default has for some reason changed then we still get the same result as that point of time (hopefully)

Reply:
Btw perhaps I'm being a noob but the example in this docstring is mighty confusing because `parse_filter_defaults` is an existing method in validphys. Is this on purpose?